diff --git a/views_hacks/views_filters_selective/views_handler_filter_selective.inc b/views_hacks/views_filters_selective/views_handler_filter_selective.inc
index 696c0c0..75f06fe 100644
--- a/views_hacks/views_filters_selective/views_handler_filter_selective.inc
+++ b/views_hacks/views_filters_selective/views_handler_filter_selective.inc
@@ -15,6 +15,21 @@ class views_handler_filter_selective extends views_handler_filter_in_operator {
         $this->value_options = array_combine(array_map('urlencode', array_values($options)), array_values($options));
       }
     }
+    // Add lost tid titles
+    $keys = array_keys($this->value_options);
+    foreach ($keys as $k => $v) {
+      if (!preg_match('/^\d+$/', $v)) {
+        unset($keys[$k]);
+      }
+    }
+    $result = db_query('SELECT tid, name FROM {taxonomy_term_data} WHERE tid IN ('.implode(',', $keys).')');
+    foreach ($result as $term) {
+      if ($this->value_options[$term->tid]) {
+        $this->value_options[$term->tid] = $term->name;
+      }
+    }
+    // Strange empty value - remove
+    unset($this->value_options['']);
   }
 
   function value_form(&$form, &$form_state) {
